tests/modules/images/regenerate-existing-images/background-process/perflab-background-job-test.php

```diff
@@ -0,0 +1,154 @@
+	 */
+	public function test_set_status_false_for_invalid_status() {
+		$result = $this->job->set_status( 'invalid_status' );
+		$this->assertFalse( $result );
+	}
+
+	public function test_create_job() {
+		$this->assertInstanceOf( Perflab_Background_Job::class, $this->job );
+	}
+
+	/**
+	 * @covers ::should_run
+	 */
+	public function test_job_should_run_for_job() {
+		$run = $this->job->should_run();
+		$this->assertTrue( $run );
+	}
+
+	/**
+	 * @covers ::should_run
+	 */
+	public function test_job_should_not_run_for_completed_job() {
+		// Mark job as complete.
+		$this->job->set_status( 'completed' );
+		$run = $this->job->should_run();
+
+		$this->assertFalse( $run );
+	}
+
+	/**
+	 * @covers ::lock
+	 * @covers ::unlock
+	 */
+	public function test_lock_unlock() {
+		$time = time();
+		$this->job->lock( $time );
+		$lock_time = get_term_meta( $this->job->get_id(), 'perflab_job_lock', true );
+		$this->assertSame( absint( $lock_time ), $time );
+		$this->job->unlock();
+		$lock_time = get_term_meta( $this->job->get_id(), 'perflab_job_lock', true );
+		$this->assertEmpty( $lock_time );
+	}
+
+	/**
+	 * @covers ::get_start_time
+	 */
+	public function test_get_start_time() {
+		$time = time();
+		$this->job->lock( $time );
+		$lock_time = get_term_meta( $this->job->get_id(), 'perflab_job_lock', true );
+		$this->assertSame( absint( $lock_time ), $time );
+
+		$start_time = $this->job->get_start_time();
+		$this->assertEquals( $start_time, $lock_time );
+		$this->assertEquals( $start_time, $time );
+		$this->assertEquals( $time, $lock_time );
+	}
+
+	public function test_max_attempts_filter() {
+		$filter_ran = false;
+		add_filter(
+			'perflab_job_max_attempts_allowed',
+			function( $value ) use ( &$filter_ran ) {
+				$filter_ran = true;
+				return $value;
+			}
+		);
+		$this->job->should_run();
+
+		$this->assertTrue( $filter_ran );
+	}
+
+	public function test_max_attempts_limit_reached() {
+		$before_attempts_limit = $this->job->should_run();
+		update_term_meta( $this->job->get_id(), $this->job::META_KEY_JOB_ATTEMPTS, 2 );
+		add_filter(
+			'perflab_job_max_attempts_allowed',
+			function() {
+				return 1;
+			}
+		);
+		$after_attempts_limit = $this->job->should_run();
+
+		$this->assertTrue( $before_attempts_limit );
+		$this->assertFalse( $after_attempts_limit );
+	}
+
+	/**
+	 * @covers ::set_status
+	 * @dataProvider job_provider
+	 */
+	public function test_set_status_false_for_valid_status( $status ) {
+		$this->assertTrue( $status );
+		$this->assertTrue( $status );
+		$this->assertTrue( $status );
+		$this->assertTrue( $status );
+	}
+
+	public function job_provider() {
+		$job = perflab_create_background_job( 'test' );
+		return array(
+			'running status'   => array( $job->set_status( 'running' ) ),
+			'partial status'   => array( $job->set_status( 'partial' ) ),
+			'failed status'    => array( $job->set_status( 'failed' ) ),
+			'completed status' => array( $job->set_status( 'completed' ) ),
+		);
```

Comment:
This data provider shouldn't create a job, that should still be part of either the test method or the overall `setUp()` method. The provider should return nothing but the simple `$status` parameter, not the result of `$job->set_status`.

Reply:
@felixarntz data provider method is called before `setUp` and `setUpBeforeClass` methods, so it won't be feasible to use `$this->job` directly.
Let me know if there is any workaround.

Reply:
@ankitrox I struggled to understand how this test worked at first glance. I feel using the data provider over complicates things in this instance when the following seems simpler and more readable?

```php
/**
 * @covers ::set_status
 */
public function test_set_status_true_for_valid_status( $status ) {
	$this->assertTrue( job->set_status( 'running' ) );
	$this->assertTrue( job->set_status( 'partial' ) );
	$this->assertTrue( job->set_status( 'failed' ) );
	$this->assertTrue( job->set_status( 'completed' ) );
}
```

Also notice the updated test name from `false_for_valid_status` to `true_for_valid_status` 

If this makes sense, can you update?

Reply:
@ankitrox @jjgrainger I think we may still have a misunderstanding here. What I am suggesting is:
```suggestion
		return array(
			'running status'   => array( 'running' ),
			'partial status'   => array( 'partial' ),
			'failed status'    => array( 'failed' ),
			'completed status' => array( 'completed' ),
		);
```
So no need to have access to or create any job instance here. Then the test method could simply be:
```php
$job = perflab_create_background_job( 'test' );
$this->assertTrue( $job->set_status( $status ) );
```

This way we have a clean use of data provider, only providing the part of the test that actually differs between test cases. In the above code snippet, we can then potentially even remove the first line, if we go with a `setUp()` method that creates `$this->job` instead. So then the test would become a single line:
```php
$this->assertTrue( $this->job->set_status( $status ) );
```

Does that make sense?

Reply:
Please see https://github.com/WordPress/performance/pull/507/files#r984782431 on what I meant here. The data provider doesn't need any job instance.